src/r3/atomic-swap/evm/src/SwapVault.sol

```diff
@@ -0,0 +1,185 @@
+
+/******************************************************************************
+ * Copyright 2023 R3 LLC                                                      *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License");            *
+ * you may not use this file except in compliance with the License.           *
+ * You may obtain a copy of the License at                                    *
+ *                                                                            *
+ *     http://www.apache.org/licenses/LICENSE-2.0                             *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ ******************************************************************************/
+
+pragma solidity 0.8.20;
+
+import "openzeppelin/token/ERC20/IERC20.sol";
+import "openzeppelin/token/ERC20/extensions/IERC20Metadata.sol";
+import "openzeppelin/token/ERC20/utils/SafeERC20.sol";
+import "openzeppelin/token/ERC721/IERC721.sol";
+import "openzeppelin/token/ERC1155/IERC1155.sol";
+import "openzeppelin/token/ERC721/utils/ERC721Holder.sol";
+import "openzeppelin/token/ERC1155/utils/ERC1155Holder.sol";
+
+library Utils {
+    function safeSupportsInterface(address addr, bytes4 interfaceId) external view returns (bool supported) {
+        (bool success, bytes memory result) = addr.staticcall(
+                abi.encodeWithSelector(IERC165.supportsInterface.selector, interfaceId)
+            );
+        
+        supported = success && result.length > 0 && abi.decode(result, (bool));
+    }
+}
+
+contract SwapVault is ERC721Holder, ERC1155Holder {
+
+    error ZERO_ADDRESS();
+    error COMMIT_TO_SELF();
+    error ALREADY_COMMITTED();
+    error INVALID_AMOUNT();
+    error INVALID_STATUS();
+    error INVALID_CLAIMER();
+    error UNSUPPORTED_INTERFACE();
+
+    event Commit(string indexed swapId, bytes32 holdHash);
+    event Transfer(string indexed swapId, bytes32 holdHash);
+    event Revert(string indexed swapId, bytes32 holdHash);
+
+    using Utils for address;
+
+    struct Commitment {
+        string swapId;
+        uint256 status;
+        address owner;
+        address recipient;
+        uint256 amount;
+        uint256 tokenId;
+        address tokenAddress;
+        uint256 signaturesThreshold;                
+        bytes32[] signatures;
+    }
+
+    mapping(string => Commitment) _committmentState;
+
+    function commit(string calldata swapId, address recipient, uint256 signaturesThreshold) external {
```

Comment:
Perhaps this is not necessary for the scope of Harmonia, but perhaps you can document the functions inside the contract like open zeppelin does [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol), This would then allow you to generate documentation automatically using tools like [solidity docgen](https://github.com/OpenZeppelin/solidity-docgen)

Reply:
@ILoomans sure, that is a good point. At the moment code will keep evolving in 2-3 more iterations probably, so best to wait for the final polish